getnacelle/nacelle-js
# [packages/shopify-checkout] Add type declarations & source maps

### WHY are these changes introduced?

#### GitHub Issue

Fixes https://github.com/getnacelle/nacelle-js/issues/18.



#### Ticket

[LD-1180](https://nacelle.atlassian.net/browse/LD-1180)

> `@nacelle/shopify-checkout` doesn't include type declarations

### WHAT is this pull request doing?



#### `@nacelle/shopify-checkout`

- Creates [type declarations](https://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html) as part of the `build` step.
  - Type declarations can be found in `dist/types`.
  - Note: this requires removing `~/module/aliases` in favor of `./relative/paths` due to the TypeScript compiler not replacing module aliases with their resolved equivalents in emitted code (see https://github.com/microsoft/TypeScript/issues/10866 for more info).
- Creates source maps in `dist/` for `es`, `umd`, and `iife` outputs.
- In addition to publishing `dist/`, publish `src/` in keeping with [best practices](https://kamyshov.info/post/shipping-type-declarations-is-not-enough/) surrounding source maps that get used with a browser's debugger.
  - The only apparent drawback is that it will increase the “Unpacked size” metric on npm. But it won't actually increase the production size of projects that use `@nacelle/shopify-checkout`, since the source maps and linked sources are often disabled in production by bundlers (like Webpack), and source maps are only loaded if the Developer Tools are open.
- Other miscellaneous improvements, such as removing the dependency on `@nacelle/types` for TypeScript users, and improved test coverage.

<details>
  <summary>Click to expand build outputs</summary>
  <img width="380" alt="dist/ containing type declarations and source maps" src="https://user-images.githubusercontent.com/5732000/135930306-9ac675bd-093e-4722-950b-a60a198117bf.png">
</details>

#### `packages/*`

- Use the latest [`vite`](https://github.com/vitejs/vite) to take advantage of [an `es` output bug fix that landed in `2.6.0-beta.3`](https://github.com/vitejs/vite/commit/06d86e4a2e90ca916a43d450bca1e6c28bc4bfe2)
  - This will result in slightly larger `.es` outputs; this should be justified by the improved debugging experience that the bug fix provides.

## 

### How to Test

1. Check out the `LD-1180-nacelle-shopify-checkout-type-defs` branch / `gh pr checkout 19`
2. `npm run bootstrap` from the monorepo root
3. Either `npm run build` from the monorepo root, or `cd packages/shopify-checkout && npm run build`
4. `npm link` from `packages/shopify-checkout`
5. In a test project, import and initialize the Shopify checkout client
  a. Quick start from the terminal on Mac OS with `npm 7`:  
```
cd ~/Desktop && mkdir checkout-example && cd checkout-example && npm init --y && npm link @nacelle/shopify-checkout && code .
```
  b. Initialize the client with:
```js
import createShopifyCheckoutClient from "@nacelle/shopify-checkout";

const checkoutClient = createShopifyCheckoutClient({
  storefrontCheckoutToken: "c913cedf66384d2e556115cc99f5d426",
  myshopifyDomain: "gamma-nova-jewelery",
  storefrontApiVersion: "2021-07"
});
```

6. Check that IntelliSense works as expected:

https://user-images.githubusercontent.com/5732000/135925892-f643cc4e-d878-491d-808d-39c67024fb9c.mov
